vortex-file/src/writer.rs

```diff
@@ -78,30 +78,29 @@ impl VortexWriteOptions {
 
         // Our buffered message writer accumulates messages for each batch so we can flush them
         // into the file.
-        let mut segment_writer = BufferedSegmentWriter::default();
-        let mut segment_specs = vec![];
-
-        // Then write the stream via the root layout
-        while let Some(chunk) = stream.next().await {
-            let chunk = chunk?;
-            layout_writer.push_chunk(&mut segment_writer, chunk)?;
-            // NOTE(ngates): we could spawn this task and continue to compress the next chunk.
-            segment_writer
-                .flush_async(&mut write, &mut segment_specs)
-                .await?;
-        }
-
-        // Flush the final layout messages into the file
-        layout_writer.flush(&mut segment_writer)?;
-        segment_writer
-            .flush_async(&mut write, &mut segment_specs)
-            .await?;
-
-        // Finish the layouts and flush the finishing messages into the file
-        let layout = layout_writer.finish(&mut segment_writer)?;
-        segment_writer
-            .flush_async(&mut write, &mut segment_specs)
-            .await?;
+        let (mut segment_buffer, segment_writer) = BufferedSegmentWriter::create();
+
+        let flush = async { segment_writer.write(write).await };
+        let write_chunks = async {
+            // Then write the stream via the root layout
+            while let Some(chunk) = stream.next().await {
+                let chunk = chunk?;
+                layout_writer.push_chunk(&mut segment_buffer, chunk)?;
+                segment_buffer.flush()?;
+            }
+
+            // Flush the final layout messages into the file
+            layout_writer.flush(&mut segment_buffer)?;
+            segment_buffer.flush()?;
+
+            // Finish the layouts and flush the finishing messages into the file
+            let layout = layout_writer.finish(&mut segment_buffer)?;
+            segment_buffer.flush()?;
+            drop(segment_buffer);
+            Ok(layout)
+        };
+        // compress chunks and flush concurrently, former is cpu bound latter is io.
+        let (layout, (mut write, segment_specs)) = future::try_join(write_chunks, flush).await?;
```

Comment:
Given that compressing chunks is CPU bound, is a concurrent wait for `write_chunks, flush` the right mechanism opposed to putting the CPU bound work on a diff thread? Say a task is currently busy compressing, this also means that no progress can be made on issuing write, right?

Reply:
Yea this assumes we don't block that much while polling compression. My thinking was that compressing an average chunk would be shorter compared to writing the previous compressed chunk into remote storage, so the combined future would have some await points during the write.

Spawning here would be better and I have to add that while doing the next thing I plan to do (multithreaded compression), but that requires changing the `VortexWrite` a bit, as to spawn we need it to be `Send + 'static`